shared/evmrpc/tx_price_oracle.go

```diff
@@ -0,0 +1,72 @@
+package evmrpc
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/skip-mev/go-fast-solver/shared/clients/coingecko"
+)
+
+const (
+	coingeckoEthID       = "ethereum"
+	coingeckoUSDCurrency = "usd"
+)
+
+// Oracle is a evm uusdc tx execution price oracle that determines the price of
+// executing a tx on chain in uusdc.
+type Oracle struct {
+	coingecko coingecko.PriceClient
+}
+
+// NewOracle creates a new evm uusdc tx execution price oracle.
+func NewOracle(coingecko coingecko.PriceClient) *Oracle {
+	return &Oracle{coingecko: coingecko}
+}
+
+// TxFeeUUSDC estimates what the cost in uusdc would be to execute a tx. The
+// tx's gas fee cap and gas limit must be set.
+func (o *Oracle) TxFeeUUSDC(ctx context.Context, tx *types.Transaction) (*big.Int, error) {
+	if tx.Type() != types.DynamicFeeTxType {
+		return nil, fmt.Errorf("tx type must be dynamic fee tx, got %d", tx.Type())
+	}
+
+	// for a dry ran tx, GasFeeCap() will be the suggested gas tip cap + base
+	// fee of current chain head
+	estimatedPricePerGas := tx.GasFeeCap()
+	if estimatedPricePerGas == nil {
+		return nil, fmt.Errorf("tx's gas fee cap must be set")
+	}
+
+	// for a dry ran tx, Gas() will be the result of calling eth_estimateGas
+	estimatedGasUsed := tx.Gas()
+	return o.gasCostUUSDC(ctx, estimatedPricePerGas, big.NewInt(int64(estimatedGasUsed)))
+}
+
+// gasCostUUSDC converts an amount of gas and the price per gas in gwei to
+// uusdc based on the current CoinGecko price of ethereum in usd.
+func (o *Oracle) gasCostUUSDC(ctx context.Context, pricePerGasGwei *big.Int, gasUsed *big.Int) (*big.Int, error) {
```

Comment:
not in love with the solution for converting the gas cost to uusdc since it involved floating point numbers which was causing some weird rounding issues where we are sometimes off by +- 1 uusdc. I tried for a bit to come up with a way to do the conversion without floats but eventually gave up lol, I feel like since it pretty much is just +- 1 uusdc, it shouldn't really make a material difference in the scheme of things since most settlements will be batches in the 1000's of usdc

Reply:
// gasCostUUSDC converts an amount of gas and the price per gas in gwei to
// uusdc based on the current CoinGecko price of ethereum in usd.
func (o *Oracle) gasCostUUSDC(ctx context.Context, pricePerGasGwei *big.Int, gasUsed *big.Int) (*big.Int, error) {
// Calculate transaction fee in Gwei
txFeeGwei := new(big.Int).Mul(gasUsed, pricePerGasGwei)
fmt.Println("tx fee gwei", txFeeGwei.String())

	// Get the ETH price in USD cents from CoinGecko
	ethPriceUSD, err := o.coingecko.GetSimplePrice(ctx, coingeckoEthID, coingeckoUSDCurrency)
	if err != nil {
		return nil, fmt.Errorf("getting CoinGecko price of Ethereum in USD: %w", err)
	}

	// Convert ETH price to microunits of USDC (uusdc) per Gwei
	// GWEI_PER_ETH = 1_000_000_000 (1 ETH = 10^9 Gwei)
	// UUSDC_PER_USD = 1_000_000 (1 USD = 10^6 UUSDC)
	const GWEI_PER_ETH = 1_000_000_000
	const UUSDC_PER_USD = 1_000_000

	// ethPriceUUSDCPerGwei = (ethPriceUSD * UUSDC_PER_USD) / GWEI_PER_ETH
	ethPriceUUSDCPerGwei := new(big.Int).Mul(big.NewInt(int64(ethPriceUSD)), big.NewInt(UUSDC_PER_USD))
	ethPriceUUSDCPerGwei.Div(ethPriceUUSDCPerGwei, big.NewInt(GWEI_PER_ETH))

	// Calculate transaction fee in uusdc
	txFeeUUSDC := new(big.Int).Mul(txFeeGwei, ethPriceUUSDCPerGwei)

	return txFeeUUSDC, nil
}

Reply:
does something like this work?

Reply:
almost, `ethPriceUSD` comes back from the coin gecko client as a `float64`, so we need to do a conversion there, but since we first converting to uusdc, that will eliminate all of the decimals so after that then we can treat it as an int which should work. TY!

Reply:
ah and this line
```
ethPriceUUSDCPerGwei.Div(ethPriceUUSDCPerGwei, big.NewInt(GWEI_PER_ETH))
```
would have to be treated as a float, we need the decimal places here

Reply:
we could do something like this
```
	tmp := new(big.Int).Mul(ethPriceUUSDCInt, txFeeGwei)
	txFeeUUSDC := new(big.Int).Div(tmp, big.NewInt(GWEI_PER_ETH))
```
but this only works if tmp has more decimals than GWEI_PER_ETH, if not then we are going to truncate some on the division

Reply:
ok pushed a change with some comments about tradeoffs here https://github.com/skip-mev/skip-go-fast-solver/pull/3/commits/9e187b89c6af440e3cf472b994a1cfc1ee1a4327, lmk what you think

Reply:
yeah that makes sense to me 👍